### ISSUE TITLE
What does MutableScope mean?

### ISSUE THREAD

**Issue by marsggbo** [microsoft/nni#1935]
what does `mutables.MutableScope`  mean? when should we use this class? I see that it is only used in ENAS example, but `enter_mutable_scope` method is not overrided. Could you explain it, thx!

**Comment by QuanluZhang** [comment]
Thanks @marsggbo for asking. The detailed document for the design of NNI NAS interface and how to write a NAS algorithm on NNI is still ongoing, and will be complete in release v1.4

For your question, we introduce `MutableScope` in the interface because some NAS algorithms require structural/hierarchical information, for example, which `LayerChoice`s and `InputChoice` are within a cell. So a `MutableScope` is like a `tree`, it tells which `LayerChoice`, `InputChoice`, `MutableScope` are under it.

For ENAS controller, it adds one more time step in every boundary of cells, whose hidden state represents the corresponding cell. Some NAS algorithms do not care about the structural information, then they can simply ignore them.

For your last question, @ultmaster will give you detailed answer.

**Comment by ultmaster** [comment]
For this *extra step*, I think it's not mentioned in the paper, you can refer to [enas code](https://github.com/melodyguan/enas/blob/master/src/cifar10/micro_controller.py#L198) for details.

Writing your own mutator is experimental and not written in docs yet, but FYI, to implement a new mutator, there are two ways:

1. Inherit `BaseMutator` and implement a bunch of callbacks, like `on_forward_layer_choice`, `enter_mutable_scope`, each will be triggered at some place during the forward place. This is the way #1863 adopted. As the control flow is implemented in a callback manner, **this is for advanced users only**. For your question, `enter_mutable_scope` and `exit_mutable_scope` is triggered on the enter and exit event of mutable scope. Implement them if your mutator can leverage that knowledge.
2. Inherit `Mutator` and implement **only** `sample_search` and `sample_final`, which corresponds to sampling an architecture for search, and reporting an architecture for export, respectively. In this case, you need to implement **none of the callback functions**. In this case, forward method is automatically based on your selected architecture. As the information of mutable scope is already embedded in the tree-structured search space, you can leverage that information when traversing this tree.

TLDR: in ENAS example, the mutable scope is used to construct the structured search space, and overriding callback functions is not needed.
